docs/config_file.rst
.. _config_file:

Zoe configuration
=================

Zoe can be configured by files, environment variables or commandline options. Most of the configuration directives listed in this file can be specified by any of the three methods, with some exception outlined below. Use the ``--help`` command-line option to have more details on the format of environment variables and precedence rules.

Command-line options
--------------------

These options can be specified only on the command-line:

* ``--write-config <filename>`` : generate a configuration file with all options set to the default values
* ``--log-file <filename>`` : write the log output to the specified file instead of ``stderr``

zoe.conf
--------

The Zoe config file have a simple format of ``<option name> = <value>``. Dash characters can be use for comments.
All Zoe processes use a single configuration file, called zoe.conf. It is looked up in the current working directory and in ``/etc/zoe/``.

Common options:

* ``debug = <true|false>`` : enable or disable debug log output
* ``deployment-name = devel`` : name of this Zoe deployment. Can be used to have multiple Zoe deployments using the same back-end (devel and prod, for example)

Workspaces:

* ``workspace-deployment-path`` : path appended to the ``workspace-base-path`` to distinguish this deployment. If left unspecified it is equal to the deployment name
* ``workspace-base-path = /mnt/zoe-workspaces`` : Base directory where user workspaces will be created. This directory should reside on a shared filesystem visible by all hosts where containers will be run.
* ``fs-group-id = 5001`` : Group ID to use for all Zoe users in workspace files

Metrics:

* ``kairosdb-enable = false`` : Enable gathering of usage metrics recorded in KairosDB
* ``kairosdb-url = http://localhost:8090`` : URL of KairosDB REST API
* ``influxdb-enable = false`` : Enable gathering of usage metrics recorded in InfluxDB
* ``kairosdb-url = http://localhost:8086`` : URL of InfluxDB REST API

Service logs (see: :ref:`logging`):

* ``gelf-address`` : Enable Docker GELF log output to a UDP listener (ex. udp://1.2.3.4:7896), works only for the Swarm back-end
* ``gelf-listener = 7896`` : Enable the internal GELF log listener on this port, set to 0 to disable
* ``service-logs-base-path = /var/lib/zoe/service-logs`` : Path where service logs coming from the GELF listener will be stored

PostgresQL database options:

* ``dbname = zoe`` : DB name
* ``dbuser = zoe`` : DB user
* ``dbpass = zoe`` : DB password
* ``dbhost = localhost`` : DB hostname
* ``dbport = 5432`` : DB port

API options:

* ``listen-address`` : address Zoe will use to listen for incoming connections to the web interface
* ``listen-port`` : port Zoe will use to listen for incoming connections to the web interface
* ``master-url = tcp://127.0.0.1:4850`` : address of the Zoe Master ZeroMQ API
* ``cookie-secret = changeme``: secret used to encrypt cookies
* ``zapp-shop-path = /var/lib/zoe-apps`` : path to the directory containing the ZApp Shop files

Master options:

* ``api-listen-uri = tcp://*:4850`` : ZeroMQ server connection string, used for the master listening endpoint
* ``max-core-limit = 16`` : maximum amount of cores a user is able to reserve
* ``max-memory-limit = 64`` : maximum amount of memory a user is able to reserve
* ``additional-volumes = <none>`` : list of additional volumes to mount in every service, for every ZApp (ex. /mnt/data:data,/mnt/data_n:data_n)

Authentication:

* ``auth-file = zoepass.csv`` : Path to the CSV file containing user,pass,role lines for text authentication
* ``ldap-server-uri = ldap://localhost`` : LDAP server to use for user authentication
* ``ldap-bind-user = ou=something,dc=any,dc=local`` : LDAP user for binding to the server
* ``ldap-bind-password = mysecretpassword`` : Password for the bind user
* ``ldap-base-dn = ou=something,dc=any,dc=local`` : LDAP base DN for users

Scheduler options:

* ``scheduler-class = <ZoeElasticScheduler>`` : Scheduler class to use for scheduling ZApps (default: elastic scheduler)
* ``scheduler-policy = <FIFO | SIZE>`` : Scheduler policy to use for scheduling ZApps (default: FIFO)
* ``placement-policy = <waterfill | random | average>`` : how containers should be placed on hosts (default: average)

ZApp shop:

* ``zapp-shop-path = /var/lib/zoe-apps`` : Path where ZApp folders are stored

Back-end choice:

* ``backend = <DockerEngine|Kubernetes>`` : cluster back-end to use to run ZApps, default is DockerEngine

Kubernetes back-end:

* ``kube-config-file = /opt/zoe/kube.conf`` : the configuration file of Kubernetes cluster that zoe works with. Specified if ``backend`` is ``Kubernetes``.

DockerEngine back-end:

* ``backend-docker-config-file = docker.conf`` : name of the DockerEngine back-end configuration file
* ``overlay-network-name = zoe`` : Name of the Docker network Zoe should use (can be overridden in the ZApp definition)

Proxy options:

By default proxy support is disabled. To configure it refer to the :ref:`proxy documentation <proxy>`.


CHANGELOG.md
# Zoe Changelog

## Version 2017.12

* New Docker Engine back-end, the Swarm back-end is now deprecated
* Nodes and ZApps can be labelled for constraining execution placement, for example to run ZApps only on nodes with GPUs
* Use non-reserved memory, cores, labels and image availability to take placement decisions
* The elastic scheduler is considered stable, the simple scheduler is now deprecated
* Expand the status page for the administrator
* More information about authentications in the log output of zoe-api
* Endpoint links in the web interface open in new windows
* Distinguish between reserved, allocated and in-use resources
* Allocate cores automatically, respecting the minimum configured in the ZApp
* Small graphic updates to the execution inspect web page
* Allow for more options and resource limits to be customized on the web interface (users or admins, not for guests), maximum limits are set in the zoe.conf file
* Additional volumes can be mounted by specifying them in the zoe.conf file
* Update unit and integration testing
* Elastic services that die are rescheduled on a new node
* Optional support for gathering usage metrics via KairosDB, for now these metrics are only used in the status page plots
* Fix UTC and timezone bugs for execution timestamps
* More configuration options for LDAP authentication

## Version 2017.09

* Major web UI redesign
* New app shop, users can choose applications from a catalog in the web interface
* New status page for administrators to check on the elastic scheduler status
* Split command-line tools for user and admin tasks
* Provide an integrated way of managing ZApp logs, by receiving UDP GELF messages into a simple directory structure
* The elastic scheduler is now fully tested and ready for production use 
* Extensive documentation updates to the install procedure

## Version 2017.06

* ZApp description format revision
* use JSON schemas to validate ZApps
* Expand the execution list API, adding filtering capabilities to limit the number of results returned
* Several minor bug fixes
* Deployment scripts and zoe frontend moved to external repository for easier maintenance and testing
* Removed proxy code
* ZApps have been split into multiple repositories for easier maintenance, testing and automated building
* Tyk or Kong can be used to provide SSL termination and authentication to the API
* Documentation update
* A new script `create_db_tables.py` can be used to create the schema in an empty database, useful for CI scenarios

## Version 2017.03

* Major documentation update
* Deployment scripts to install Zoe on Linux (Kubernetes or Swarm back-ends) or Windows
* Kubernetes back-end
* Updated Swarm backend, uses the new Docker Python API
* Zoe code now goes through an automated testing pipeline
* Added unit and integration tests
* Preview of the new AngularJS front-end

## Version 0.10.2

* Documentation updates
* Log management bug fixes
* Fix a bug with the scheduler time trigger
* Sort by ID when listing execution from the command-line client

## Version 0.10.1

* Service discovery API endpoint: simple, read-only, unauthenticated access to a list of DNS names of services. Needed for frameworks that need a list of hosts for configuration, can be used by scripts in the images.
* Add `-s` option to the `start` command of the commandline client to have it wait for execution termination and stream the log meanwhile.
* Add the `workspace-deployment-path` option in the configuration file. It should be used when the workspace path should be build not with the deployment name, but with something else.
* All errors generated by docker while creating containers are now considered as fatal, except the "not enough free resource" one. When a fatal error is generated, Zoe will not try to start the execution again.
* Error messages generated by Docker are now exposed to the user
* (experimental) if the configuration option `service-log-path` is set, container output is saved from docker into the specified directory. If the logs are big, this can have a significant impact on execution termination time

## Version 0.10.0

* This version is the start of a new series of releases moving toward a new architecture
* Only one overlay network (created by the sysadmin) will be used by all Zoe executions. We provide Dockerfiles for SOCKS and SSHd containers that can be used to give users access inside the Zoe overlay network. These containers will no longer be managed by Zoe.
* Move the REST API in the web process and add a ZeroMQ-based API between the web and the master.
* Move all user management in the web process.
* Use Postgresql to store the state
* Rename the zoe-web component in zoe-api to better describe its new role
* User authentication can be performed by a CSV text file or via LDAP

## Version 0.9.7

* Check application description version during validation.
* Bump application description version to 2 since we are going to make some major changes in the format
* Add fields `total_count`, `essential_count` and `startup_order` to service descriptions.
* Comment code related to the parsing of cluster statistics from Swarm. The data returned by API changes too often and parsing it consistently is too complex. A bug is open on the Swarm issue tracker.

## Version 0.9.6

* Workspaces: Zoe now supports starting containers with a directory from the hosts mounted as a volume. The directory is private for the user and is not wiped when the Zoe execution terminates. There are some issues with file permissions when the container images define users with arbitrary UIDs.
* Add an error status and error message visible to the user when Zoe fails starting an execution
* Add a Docker Compose file to help test deployments
* The gateway container image is now configurable
* Move applications to their own repository (zoe-applications)
* Move the Zoe logger code into its own repository

zoe_lib/config.py
# Copyright (c) 2015, Daniele Venzano
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Configuration parsing."""

import logging

from zoe_lib.configargparse import ArgumentParser, Namespace

logging.getLogger('requests').setLevel(logging.WARNING)
logging.getLogger('urllib3').setLevel(logging.WARNING)
logging.getLogger('docker').setLevel(logging.INFO)

log = logging.getLogger(__name__)

_CONFIG_PATHS = [
    'zoe.conf',
    '/etc/zoe/zoe.conf'
]

_CONF = None


def get_conf() -> Namespace:
    """Returns the conf singleton."""
    return _CONF


def load_configuration(test_conf=None):
    """Parses command line arguments."""
    global _CONF
    if test_conf is None:
        argparser = ArgumentParser(description="Zoe - Container Analytics as a Service",
                                   default_config_files=_CONFIG_PATHS,
                                   auto_env_var_prefix="ZOE_",
                                   args_for_setting_config_path=["--config"],
                                   args_for_writing_out_config_file=["--write-config"])

        # Common options
        argparser.add_argument('--debug', action='store_true', help='Enable debug output')
        argparser.add_argument('--deployment-name', help='name of this Zoe deployment', default='prod')

        argparser.add_argument('--dbname', help='DB name', default='zoe')
        argparser.add_argument('--dbuser', help='DB user', default='zoe')
        argparser.add_argument('--dbpass', help='DB password', default='')
        argparser.add_argument('--dbhost', help='DB hostname', default='localhost')
        argparser.add_argument('--dbport', type=int, help='DB port', default=5432)

        # Master options
        argparser.add_argument('--api-listen-uri', help='ZMQ API listen address', default='tcp://*:4850')

        argparser.add_argument('--kairosdb-enable', action="store_true", help='Enable usage metric input from KairosDB')
        argparser.add_argument('--kairosdb-url', help='URL of the KairosDB service (ex. http://localhost:8090)', default='http://localhost:8090')
        argparser.add_argument('--influxdb-enable', action="store_true", help='Enable usage metric input from InfluxDB')
        argparser.add_argument('--influxdb-url', help='URL of the InfluxDB service (ex. http://localhost:8086)', default='http://localhost:8086')

        argparser.add_argument('--workspace-base-path', help='Base directory where user workspaces will be created. Must be visible at this path on all hosts.', default='/mnt/zoe-workspaces')
        argparser.add_argument('--workspace-deployment-path', help='Path appended to the workspace path to distinguish this deployment. If unspecified is equal to the deployment name.', default='--default--')
        argparser.add_argument('--overlay-network-name', help='Name of the Swarm overlay network Zoe should use', default='zoe')

        # Service logs
        argparser.add_argument('--gelf-address', help='Enable Docker GELF log output to this destination (ex. udp://1.2.3.4:7896)', default='')
        argparser.add_argument('--gelf-listener', type=int, help='Enable the internal GELF log listener on this port, set to 0 to disable', default='7896')
        argparser.add_argument('--service-logs-base-path', help='Path where service logs coming from the GELF listener will be stored', default='/var/lib/zoe/service-logs')

        # API options
        argparser.add_argument('--listen-address', help='Address to listen to for incoming connections', default="0.0.0.0")
        argparser.add_argument('--listen-port', type=int, help='Port to listen to for incoming connections', default=5001)
        argparser.add_argument('--master-url', help='URL of the Zoe master process', default='tcp://127.0.0.1:4850')
        argparser.add_argument('--cookie-secret', help='secret used to encrypt cookies', default='changeme')

        argparser.add_argument('--auth-file', help='Path to the CSV file containing user,pass,role lines for text authentication', default='zoepass.csv')

        argparser.add_argument('--ldap-server-uri', help='LDAP server to use for authentication', default='ldap://localhost')
        argparser.add_argument('--ldap-bind-user', help='Full LDAP user to use for binding', default='ou=something,dc=any,dc=local')
        argparser.add_argument('--ldap-bind-password', help='Password for the bind user', default='mysecretpassword')
        argparser.add_argument('--ldap-base-dn', help='LDAP base DN for users', default='ou=something,dc=any,dc=local')

        argparser.add_argument('--oauth-client-id', help='OAuth2 client ID as generated by your identity provider')
        argparser.add_argument('--oauth-client-secret', help='OAuth2 client secret as generated by your identity provider')
        argparser.add_argument('--oauth-redirect-uri', help='Full URL of the Zoe API OAuth callback', default='https://my.zoe.com/api/v7/user/oauth')
        argparser.add_argument('--oauth-role', help='Role to assign to new users authenticated via OAuth2', default='user')
        argparser.add_argument('--oauth-quota', help='Quota to assign to new users authenticated via OAuth2', default='default')
        argparser.add_argument('--oauth-create-workspace-script', help='Full path to a script that creates user workspace, Zoe will call using sudo and pass username and fs_id as arguments', default='/usr/local/bin/zoe_create_workspace.sh')

        argparser.add_argument('--fs-group-id', type=int, help='Group ID to use for all Zoe users in workspace files', default='5001')

        # Proxy options
        argparser.add_argument('--proxy-path', help='Proxy base path', default='127.0.0.1')
        argparser.add_argument('--reverse-proxy-path', help='Base path in case Zoe is behind a reverse proxy under a path', default='')
        argparser.add_argument('--websocket_base', help='Base URL for websocket connections, you need to change it only when running Zoe behind a reverse proxy', default='ws://{{ server_address }}')
        argparser.add_argument('--traefik-zk-ips', help='ZooKeeper address storing dynamic configuration for træfik', default=None)
        argparser.add_argument('--traefik-base-url', help='Base path used in reverse proxy URLs generated for træfik', default='/zoe/proxy/')

        # Scheduler
        argparser.add_argument('--scheduler-class', help='Scheduler class to use for scheduling ZApps', choices=['ZoeElasticScheduler'], default='ZoeElasticScheduler')
        argparser.add_argument('--scheduler-policy', help='Scheduler policy to use for scheduling ZApps', choices=['FIFO', 'SIZE', 'DYNSIZE'], default='FIFO')
        argparser.add_argument('--placement-policy', help='Placement policy', choices=['waterfill', 'random', 'average'], default='average')

        argparser.add_argument('--backend', choices=['Kubernetes', 'DockerEngine'], default='DockerEngine', help='Which backend to enable')

        # Docker Engine backend options
        argparser.add_argument('--backend-docker-config-file', help='Location of the Docker Engine config file', default='docker.conf')

        # Kubernetes backend
        argparser.add_argument('--kube-config-file', help='Kubernetes configuration file', default='/opt/zoe/kube.conf')
        argparser.add_argument('--kube-namespace', help='The namespace that Zoe operates on', default='default')

        # other options
        argparser.add_argument('--zapp-shop-path', help='Path where ZApp folders are stored', default='/var/lib/zoe-apps')
        argparser.add_argument('--log-file', help='output logs to a file', default='stderr')
        argparser.add_argument('--max-core-limit', help='Maximum amount of cores users are able to reserve', type=int, default=16)
        argparser.add_argument('--max-memory-limit', help='Maximum amount of memory services can use (in GiB)', type=int, default=64)
        argparser.add_argument('--additional-volumes', help='Additional volumes to mount in services filesystems. (ex: /mnt/data:data,/mnt/data_n:data_n)', default='')
        argparser.add_argument('--enable-plots', action='store_true', help='Enable generation of URLs to Grafana')
        argparser.add_argument('--enable-cephfs-quotas', action='store_true', help='Enable reading cephfs quotas (needs sudo configuration)')

        opts = argparser.parse_args()
        if opts.debug:
            argparser.print_values()

        if opts.workspace_deployment_path == '--default--':
            opts.workspace_deployment_path = opts.deployment_name

        if len(opts.additional_volumes) > 0:
            vols = str(opts.additional_volumes).split(',')
            opts.additional_volumes = [v.split(':') for v in vols]
            log.info('Additional volumes:')
            for path, mountpoint in opts.additional_volumes:
                log.info('  - {} -> {}'.format(path, mountpoint))
        else:
            opts.additional_volumes = []

        _CONF = opts
    else:
        _CONF = test_conf
